libs/@local/hashql/core/src/type/inference/solver/test.rs

```diff
@@ -1579,3 +1595,334 @@ fn projection_unify_variables_equal_is_not_equivalent() {
         TypeCheckDiagnosticCategory::ConflictingEqualityConstraints
     );
 }
+
+#[test]
+fn single_subscript() {
+    // T = Number[]
+    // T[0]
+    let heap = Heap::new();
+    let env = Environment::new(SpanId::SYNTHETIC, &heap);
+
+    let list = list!(env, primitive!(env, PrimitiveType::Number));
+
+    let mut environment = InferenceEnvironment::new(&env);
+    let variable = environment.add_subscript(
+        SpanId::SYNTHETIC,
+        list,
+        primitive!(env, PrimitiveType::Integer),
+    );
+
+    let solver = environment.into_solver();
+    let (substitution, diagnostics) = solver.solve();
+    assert!(diagnostics.is_empty());
+
+    assert_equiv!(
+        env,
+        [substitution
+            .infer(variable.kind.hole().expect("variable should be hole"))
+            .expect("should have inferred variable")],
+        [union!(
+            env,
+            [
+                primitive!(env, PrimitiveType::Null),
+                primitive!(env, PrimitiveType::Number)
+            ]
+        )]
+    );
+}
+
+#[test]
+#[ignore = "https://linear.app/hash/issue/H-4645/implement-subscript-for-intersections"]
+fn multi_subscript() {
+    // given:
+    // T = List<List<String>>
+    // T[0]?[0]
+    let heap = Heap::new();
+    let env = Environment::new(SpanId::SYNTHETIC, &heap);
+
+    let list = list!(env, list!(env, primitive!(env, PrimitiveType::String)));
+
+    let mut environment = InferenceEnvironment::new(&env);
+
+    let first = environment.add_subscript(
+        SpanId::SYNTHETIC,
+        list,
+        primitive!(env, PrimitiveType::Integer),
+    );
+
+    // remove `Null` from the first union
+    let second = environment.add_subscript(
+        SpanId::SYNTHETIC,
+        intersection!(
+            env,
+            [
+                first.into_type(&env).id,
+                list!(env, primitive!(env, PrimitiveType::String))
+            ]
+        ),
+        primitive!(env, PrimitiveType::Integer),
+    );
+
+    let solver = environment.into_solver();
+    let (substitution, diagnostics) = solver.solve();
+    assert!(diagnostics.is_empty());
+
+    assert_equiv!(
+        env,
+        [substitution
+            .infer(second.kind.hole().expect("variable should be hole"))
+            .expect("should have inferred variable")],
+        [union!(
+            env,
+            [
+                primitive!(env, PrimitiveType::String),
+                primitive!(env, PrimitiveType::Null)
+            ]
+        )]
+    );
+}
+
+#[test]
+fn early_subscript() {
```

Comment:
Shouldn't this be `late_subscript` and the other one `early_subscript`?

Reply:
The name pertains to the fact that the inference variable has been added to the constraint set *before* we add the subscript. So it's early in the sense that we already know the type at this point, even if we weren't using fix-point evaluation.